include/deal.II/lac/vector.templates.h

```diff
@@ -614,6 +563,487 @@ namespace internal
+        // recursively. Make pieces (except last) divisible by 1024.
+        const size_type new_size = (vec_size / 4096) * 1024;
+        ResultType r0, r1, r2, r3;
+        accumulate_recursive (op, first, first+new_size, r0);
+        accumulate_recursive (op, first+new_size, first+2*new_size, r1);
+        accumulate_recursive (op, first+2*new_size, first+3*new_size, r2);
+        accumulate_recursive (op, first+3*new_size, last, r3);
+        r0 += r1;
+        r2 += r3;
+        result = r0 + r2;
+      }
+  }
+
+
+
+#ifdef DEAL_II_WITH_THREADS
+  /**
+   * This struct takes the loop range from the tbb parallel for loop and
+   * translates it to the actual ranges of the reduction loop inside the
+   * vector. It encodes the grain size but might choose larger values of
+   * chunks than the minimum grain size. The minimum grain size given to tbb
+   * is then simple 1. For affinity reasons, the layout in this loop must be
+   * kept in sync with the respective class for plain for loops further up
+   */
+  template <typename Operation, typename ResultType>
+  struct TBBReduceFunctor
+  {
+    TBBReduceFunctor(const Operation   &op,
+                     const size_type    vec_size)
+      :
+      op(op),
+      vec_size(vec_size)
+    {
+      // set chunk size for sub-tasks
+      const unsigned int gs = internal::Vector::minimum_parallel_grain_size;
+      n_chunks = std::min(4*MultithreadInfo::n_threads(), vec_size / gs);
+      chunk_size = vec_size / n_chunks;
+
+      // round to next multiple of 512 (or leave it at the minimum grain size
+      // if that happens to be smaller)
+      if (chunk_size > 512)
+        chunk_size = ((chunk_size + 511)/512)*512;
+      n_chunks = (vec_size + chunk_size - 1) / chunk_size;
+      AssertIndexRange((n_chunks-1)*chunk_size, vec_size);
+      AssertIndexRange(vec_size, n_chunks*chunk_size+1);
+
+      if (n_chunks > 512)
+        {
+          large_array.resize(n_chunks);
+          array_ptr = &large_array[0];
+        }
+      else
+        array_ptr = &small_array[0];
+    };
+
+    void operator() (const tbb::blocked_range<size_type> &range) const
+    {
+      for (size_type i = range.begin(); i < range.end(); ++i)
+        accumulate_recursive(op, i*chunk_size, std::min((i+1)*chunk_size, vec_size),
+                             array_ptr[i]);
+    }
+
+    ResultType do_sum() const
+    {
+      while (n_chunks > 1)
+        {
+          if (n_chunks % 2 == 1)
+            array_ptr[n_chunks++] = ResultType();
+          for (size_type i=0; i<n_chunks; i+=2)
+            array_ptr[i/2] = array_ptr[i] + array_ptr[i+1];
+          n_chunks /= 2;
+        }
+      return array_ptr[0];
+    }
+
+    const Operation &op;
+    const size_type vec_size;
+
+    mutable unsigned int n_chunks;
+    unsigned int chunk_size;
+    ResultType small_array [512];
+    std::vector<ResultType> large_array;
+    mutable ResultType *array_ptr;
```

Comment:
Can you say here that the variable either points into the small or large array, depending on the size of the vector to be operated on?